### PR TITLE
feat: add asyncapi 3.0

### PR DESCRIPTION
**Pull request by AlexVarchuk**
## What/Why/How?
Add supporting asyncAPI 3.0
## Check yourself

- [ ] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [ ] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/config/__tests__/__snapshots__/config.test.ts.snap [hunk 1]
```diff
@@ -12,6 +12,11 @@ StyleguideConfig {
       "oas3_0": {},
       "oas3_1": {},
     },
+    "async3": {
+      "oas2": {},
+      "oas3_0": {},
+      "oas3_1": {},
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/config/__tests__/__snapshots__/config.test.ts.snap:18]: the reason of that is populating all rules in config.

File: packages/core/src/config/builtIn.ts [hunk 2]
```diff
@@ -33,18 +34,21 @@ export const defaultPlugin: Plugin = {
...
     oas2: oas2Rules,
     async2: async2Rules,
+    async3: async3Rules,
     arazzo: arazzoRules,
   } as CustomRulesConfig,
   preprocessors: {
     oas3: oas3Preprocessors,
     oas2: oas2Preprocessors,
     async2: async2Preprocessors,
+    async3: {},
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/config/builtIn.ts]: done

File: packages/core/src/types/asyncapi2.ts [hunk 3]
```diff
@@ -410,6 +410,9 @@ const Schema: NodeType = {
     additionalProperties: (value: any) => {
       return typeof value === 'boolean' ? { type: 'boolean' } : 'Schema';
     },
+    additionalItems: (value: any) => {
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi2.ts]: removed. Added only in 3.0

**Review by jeremyfiel — COMMENTED** [review]
hi @AlexVarchuk 

Please let me know if you have any questions

File: packages/core/src/types/asyncapi3.ts [hunk 4]
```diff
@@ -0,0 +1,1189 @@
...
+const MessageExample: NodeType = {
+  properties: {
+    payload: { isExample: true },
+    summary: { type: 'string' },
+    name: { type: 'string' },
+    headers: { type: 'object' },
+  },
+};
+
+const Schema: NodeType = {
```
Comment (tatomyr) [inline packages/core/src/types/asyncapi3.ts:371]: Could you reuse the 2.6 schema as it shares the [same draft](https://www.asyncapi.com/docs/reference/specification/v3.0.0#schemaObject) with 3.0?

File: packages/core/src/typings/asyncapi3.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
+export interface Async3Definition {
+  asyncapi: string;
+  info?: Async3Info;
+  channels: Record<string, Channel>;
+  // TBD
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/typings/asyncapi3.ts]: added more types. and removed TBD. Thanks

File: packages/core/src/types/asyncapi3.ts [hunk 6]
```diff
@@ -0,0 +1,1189 @@
...
+  },
+  additionalProperties: {},
+};
+
+const MessageBindings: NodeType = {
+  properties: {},
+  allowed() {
+    // allow all supported values, not all have deep linting
+    return [
+      'http',
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi3.ts]: now we use if from `async2.ts`

File: packages/core/src/types/asyncapi3.ts (hunk 4 above)
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi3.ts]: done. But also fixed types for `schema->items` for async2. It is related to the `bundle` command.

File: packages/core/src/types/asyncapi3.ts [hunk 7]
```diff
@@ -0,0 +1,1189 @@
...
+  properties: {}, // empty object
+};
+MessageBindings.properties.mercure = MercureMessageBinding;
+
+const MercureOperationBinding: NodeType = {
+  properties: {}, // empty object
+};
+OperationBindings.properties.mercure = MercureOperationBinding;
+
+// ibmmq
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi3.ts]: removed. It was one of the possible types for `ChannelBindings`. Now we use type from async2.ts

**Review by JLekawa — APPROVED** [review]

File: packages/core/src/types/asyncapi2.ts [hunk 8]
```diff
@@ -1099,12 +1086,32 @@ export const AsyncApi2Types: Record<string, NodeType> = {
...
   ChannelBindings,
+  MessageBindings,
+  OperationBindings,
+};
+
+export const AsyncApi2Types: Record<string, NodeType> = {
+  ...AsyncApi2Bindings,
+  Root,
+  Tag,
+  TagList: listOf('Tag'),
```
Comment (jeremyfiel) [inline packages/core/src/types/asyncapi2.ts:1097]: should this remain consistent with other `type` files.  `NamedTags: listOf('Tag')` ??

File: packages/core/src/types/asyncapi2.ts [hunk 9]
```diff
@@ -1099,12 +1086,32 @@ export const AsyncApi2Types: Record<string, NodeType> = {
...
+  Info,
+  Contact,
+  License,
   ChannelMap,
   Channel,
   Parameter,
   ParametersMap: mapOf('Parameter'),
   Operation,
   Schema,
+  SchemaList: listOf('Schema'),
```
Comment (jeremyfiel) [inline packages/core/src/types/asyncapi2.ts:1114]: should this remain consistent with other `type` files.  `NamedSchemas: listOf('Schema')` ??

File: packages/core/src/types/asyncapi2.ts [hunk 10]
```diff
@@ -1099,12 +1086,32 @@ export const AsyncApi2Types: Record<string, NodeType> = {
...
+  Root,
+  Tag,
+  TagList: listOf('Tag'),
+  ServerMap,
+  ExternalDocs,
+  Server,
+  ServerVariable,
+  ServerVariablesMap: mapOf('ServerVariable'),
+  SecurityRequirement,
+  SecurityRequirementList: listOf('SecurityRequirement'),
```
Comment (jeremyfiel) [inline packages/core/src/types/asyncapi2.ts:1104]: should this remain consistent with other `type` files.  `NamedSecurityRequirements: listOf('SecurityRequirement')` ??

**Review by jeremyfiel — CHANGES_REQUESTED** [review]

File: packages/core/src/types/asyncapi3.ts (hunk 4 above)
Comment (jeremyfiel) [reply packages/core/src/types/asyncapi3.ts]: `Schema` definition should use #1608 updates. 

originally, asyncapi.ts had the Schema object defined with the wrong JSON Schema draft. It was using 2020-12 definitions, but it should be draft-07.

File: packages/core/src/types/asyncapi2.ts (hunk 10 above)
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi2.ts:1104]: it is consistent, I believe you miss difference between `listOf/mapOf` 😊 . It is a list of something. Please check the type below:
```
MessageTraitList: listOf('MessageTrait'),
OperationTraitList: listOf('OperationTrait'),
```

File: packages/core/src/types/asyncapi2.ts (hunk 9 above)
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi2.ts:1114]: the same thing

File: packages/core/src/types/asyncapi2.ts (hunk 8 above)
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi2.ts:1097]: same. please check answer below)

File: packages/core/src/types/asyncapi3.ts (hunk 4 above)
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi3.ts]: Thanks, @jeremyfiel, please merge your PR and I'll rebase it. I know about draft-07 and don't want to duplicate your PR, it'll produce conflict between our branches.

File: packages/core/src/types/asyncapi2.ts (hunk 10 above)
Comment (jeremyfiel) [reply packages/core/src/types/asyncapi2.ts:1104]: I mean the naming of the property. 

In OAS types files, the naming is `Named*` and listOf. 

In this file you are using  `*List`

Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi2.ts:1104]: If you find somewhere using `listOf` and constant name has pattern `Named*` please open issue or write here. I'll double-check it. 

Here is examples of using listof in OAS types:
https://github.com/Redocly/redocly-cli/blob/36b09f11d54c6f386965f5f21e2e4ca8b9be674a/packages/core/src/types/oas3.ts#L534-L550
https://github.com/Redocly/redocly-cli/blob/36b09f11d54c6f386965f5f21e2e4ca8b9be674a/packages/core/src/types/oas2.ts#L446-L479


As you could see, we use `*List` pattern for it

Comment (jeremyfiel) [reply packages/core/src/types/asyncapi2.ts:1104]: ah, my apologies.  I guess I was looking at the mapOf entries. 

disregard my comments.  thanks!

**Review by tatomyr — COMMENTED** [review]
A few more comments.

File: packages/core/src/types/asyncapi3.ts [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+  Schema,
+  Dependencies,
+  Discriminator,
+  DiscriminatorMapping,
+  SchemaProperties,
+} from './asyncapi2';
+
+const Root: NodeType = {
+  properties: {
+    asyncapi: null, // TODO: validate semver format and supported version
```
[the remark this answers is not in the document]
Comment (jeremyfiel) [reply packages/core/src/types/asyncapi3.ts:13]: A pattern is probably better for future compatibility. Any new enum will be a breaking change 

Something like 

```suggestion
    "pattern": "^3\\.\\d+\\.\\d+(-.+)?$"
```

File: packages/core/src/types/asyncapi3.ts [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+    description: { type: 'string' },
+    servers: 'ServerList',
+    parameters: 'ParametersMap',
+    bindings: 'ChannelBindings',
+    tags: 'TagList',
+    externalDocs: 'ExternalDocs',
+  },
+};
+
+const Tag: NodeType = {
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply packages/core/src/types/asyncapi3.ts]: reused

File: resources/async.yaml [hunk 13]
```diff
@@ -135,7 +135,7 @@ servers:
         default: '9092'
 
 channels:
-  smartylighting/streetlights/1/0/event/{streetlightId}/lighting/measured:
+  /smartylighting/streetlights/1/0/event/{streetlightId}/lighting/measured/:
```
[the remark this answers is not in the document]
Comment (AlexVarchuk) [reply resources/async.yaml]: removed. I tested our lint command)

File: __tests__/bundle/async3/simple.yml [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  userSignedup:
+    address: user/signedup
+    messages:
+      UserSignedUp:
+        $ref: '#/components/messages/UserSignedUp'
+operations:
+  sendUserSignedup:
+    action: send
+    channel:
+      $ref: '#/channels/userSignedup'
```
Comment (tatomyr) [inline __tests__/bundle/async3/simple.yml:16]: Out of curiosity, is it even legal to reference through `channels`? I thought it was supposed to use the `components` section for shared components.

**Review by tatomyr — APPROVED** [review]
Good job! Thanks @AlexVarchuk.

File: __tests__/bundle/async3/simple.yml (hunk 14 above)
Comment (AlexVarchuk) [reply __tests__/bundle/async3/simple.yml:16]: No strict rules about that. We found this `simple.yaml` from async API examples inside their repos, so it must work.

File: packages/core/src/types/asyncapi3.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeremyfiel) [reply packages/core/src/types/asyncapi3.ts]: this is marked resolved, but not updated in the pull request

**Comment by lornajane** [comment]
I am so happy to see this feature landing! Does this new AsyncAPI 3.0 support include bundling, or is it lint-only? There's a limitation where descriptions use `$ref` - details are in #1543, but I can't tell from the source if we tackled that in this pull request.

**Comment by AlexVarchuk** [comment]
Hi @lornajane. It includes improvements for `bundle` command. I tested it using our internal description. It doesn't cover all cases. If you have description to reproduce https://github.com/Redocly/redocly-cli/issues/1543 please send me and I'll check it.

**Comment by jeremyfiel** [comment]
closes #1352

**Pull request merged**